runner.py

```diff
@@ -0,0 +1,134 @@
+import runner_run_funcs
+
+SIM_TYPE_FED_LEARNING = "fed_learning"
+SIM_TYPE_FED_AVG_WITH_DP = "fed_avg_with_dp"
+
+DATA_GEN_TYPE_DATA_FROM_FILE = "file_data"
+DATA_GEN_TYPE_BLOB = "data_gen_blob"
+DATA_GEN_TYPE_RAND = "data_gen_rand"
+
+P_KEY_DATA_FILE_PATH = "data_file_path"
+
+P_KEY_NUM_SAMPLES = "num_samples"
+P_KEY_NUM_LABELS = "num_labels"
+P_KEY_NUM_FEATURES = "num_features"
+P_KEY_NUM_USERS = "num_users"
+
+P_KEY_NUM_ROUNDS = "num_rounds"
+P_KEY_BATCH_SIZE = "batch_size"
+P_KEY_NUM_EPOCHS = "num_epochs"
+
+
+class RunFuncAndReqParams:
+    def __init__(self, run_func, prereq_check_func):
+        self.run_func = run_func
+        self.prereq_params = prereq_check_func
+
+
+run_func_ltable = {
+    SIM_TYPE_FED_LEARNING: RunFuncAndReqParams(
+        runner_run_funcs.run_fed_learn_sim,
+        {
+            P_KEY_NUM_ROUNDS,
+            P_KEY_BATCH_SIZE,
+            P_KEY_NUM_EPOCHS,
+            P_KEY_NUM_SAMPLES,
+            P_KEY_NUM_FEATURES,
+            P_KEY_NUM_USERS,
+        },
+    ),
+    SIM_TYPE_FED_AVG_WITH_DP: RunFuncAndReqParams(
+        runner_run_funcs.run_fed_avg_with_dp, {P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES}
+    ),
+    DATA_GEN_TYPE_DATA_FROM_FILE: RunFuncAndReqParams(
+        runner_run_funcs.read_data_from_file,
+        {
+            P_KEY_NUM_SAMPLES,
+            P_KEY_NUM_LABELS,
+            P_KEY_NUM_FEATURES,
+            P_KEY_NUM_USERS,
+            P_KEY_DATA_FILE_PATH,
+        },
+    ),
+    DATA_GEN_TYPE_BLOB: RunFuncAndReqParams(
+        runner_run_funcs.run_data_gen_blob,
+        {P_KEY_NUM_SAMPLES, P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES, P_KEY_NUM_USERS},
+    ),
+    DATA_GEN_TYPE_RAND: RunFuncAndReqParams(
+        runner_run_funcs.run_data_gen_rand,
+        {P_KEY_NUM_SAMPLES, P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES, P_KEY_NUM_USERS},
+    ),
+}
+
+
+class Runner:
+    def __init__(self):
+        self._params = {}
+
+    def run(self, sim_type, data_gen_type):
+        sim_run_info = run_func_ltable[sim_type]
+        data_gen_run_info = run_func_ltable[data_gen_type]
+
+        if not self._sim_has_required_params_for_given_run_func(
+            sim_run_info.prereq_params, sim_type
+        ):
+            return
+
+        if not self._sim_has_required_params_for_given_run_func(
+            sim_run_info.prereq_params, data_gen_type
+        ):
+            return
+
+        print('Generating "{}" data...'.format(data_gen_type))
+        generated_data = data_gen_run_info.run_func(self._params)
+
+        print('Running the "{}" simulation...'.format(sim_type))
+        sim_run_info.run_func(self._params, generated_data)
+
+        print("Finished!")
+
+    def set_num_samples(self, num_samples):
```

Comment:
Let's get rid of most of these setters.  It doesn't make sense that someone would call one or two of the setters without calling all of the setters required to make the Runner get into a valid overall state.

I think it makes sense to just pass in a JSON blob of configuration to the Runner in the constructor, and then call the state validation code in `_sim_has_required_params_for_given_run_func`

Reply:
Agreed. This is probably nicer to work with.

Should I worry about providing an API to create the JSON?

Reply:
No, I don't think an API to create the JSON is necessary.